### PR TITLE
[ci] Add change blocker script

### PR DESCRIPTION
**Pull request by GregAC**
This adds a github action which checks if any changes in a PR match a list of patterns in BLOCKFILE. If there are matches the action fails which can be used to block the PR from being merged.

Anyone on the COMMITTERS list can authorize a change by adding a comment with:

CHANGE AUTHORIZED: path/to/file

To the PR. If there are multiple changes to authorize, one authorization is required per file and there is one authorization per line in the comment.

At least two committers must authorize the change to pass.

### CODE REVIEW

**Comment by GregAC** [comment]
Note this current setup isn't entirely foolproof, as the github action runs based upon the contents of the PR, so you could submit a PR removing the action entirely and it wouldn't be caught. Maybe there's a way we can prevent this (i.e. github looks at .github/workflows on master for actions to trigger on a PR) but I can't see an obvious way around it in the documentation.

**Comment by GregAC** [comment]
Now once merged the action will trigger and run based on whatever is checked into master rather than what comes in with the PR.

File: BLOCKFILE
Comment (marnovandermaas) [inline BLOCKFILE:1]: I assume that we will eventually populate this block file with all the RTL paths in the repository?

**Review by marnovandermaas — COMMENTED** [review]
Looks good to me. I am not approving this yet because we don't want to prematurely merge this.

File: BLOCKFILE
Comment (GregAC) [reply BLOCKFILE:1]: That's the plan. I don't want to add that in this PR so we can trial this in the repository before we block changes to anything (beyond this checker itself).

File: ci/scripts/check-pr-changes-allowed.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+        changes = changes_file.readlines()
+        changes = map(lambda c: c.strip(), changes)
+        changes = filter(lambda c: c != '', changes)
+
+    return changes
+
+
+def check_changes_against_blocklist(changes: list[str],
+                                    blocklist: list[str]) -> dict[str,
+                                                                  list[str]]:
```
Comment (moidx) [inline ci/scripts/check-pr-changes-allowed.py:45]: nit: whitespace alignment.

**Review by moidx — APPROVED** [review]

**Review by marnovandermaas — APPROVED** [review]
LGTM, adding this as is will not affect the RTL files yet, so I feel comfortable approving this.

**Review by msfschaffner — APPROVED** [review]

**Comment by GregAC** [comment]
Merging this now, I've tested it in my OT fork, passing run when an unblocked change is in a PR here: https://github.com/GregAC/opentitan/actions/runs/4946199875/jobs/8843950144 and failing run when a blocked change is in a PR here: https://github.com/GregAC/opentitan/actions/runs/4946058193/jobs/8843638443

**Pull request merged**